File: scripts/suno/suno_generate.js
#!/usr/bin/env node
/**
 * suno_generate.js — browser-drive Suno song generation (CDP).
 *
 * Suno proxies song *generation* through an obfuscated, rotating service-worker
 * path, so the generate request cannot be replayed as a clean API call. The
 * reliable, non-circumventing approach is to automate a real logged-in browser:
 * click "Create" like a user, then read results via the page's own session.
 *
 * This script connects to an ALREADY-RUNNING Chrome over CDP (one that is logged
 * into suno.com), drives the create page, polls the page's own /api/feed for the
 * new clips, and prints their metadata (incl. CDN audio URLs) as JSON.
 *
 * It does NOT download audio (the Python wrapper does that) and it does NOT touch
 * the obfuscated generate endpoint directly — generation happens via a UI click.
 *
 * Two ways to get a browser (no manual login required either way):
 *   1. LAUNCH (default): this script launches its OWN dedicated, isolated Chrome
 *      (own userDataDir, headless by default) and injects the engagement bot's
 *      persisted Suno session cookies. No login UI, and — crucially — it shares
 *      nothing with the posting-flow browser, so it can't cause the lock-ups /
 *      race conditions that come from contending over one browser.
 *   2. CONNECT: if cdpUrl is given, attach to an already-running Chrome instead.
 *
 * Usage:
 *   node suno_generate.js '<jsonArgs>'
 *   jsonArgs = { prompt, instrumental?, dryRun?, timeoutMs?, expectClips?,
 *                cdpUrl?, sessionFile?, userDataDir?, chromePath?, headless? }
 *
 * Output: a single JSON object on the LAST stdout line:
 *   { ok, dryRun?, clips:[{id,title,status,audioUrl,imageUrl}], error?, log:[...] }
 */

'use strict';

const fs = require('fs');
const os = require('os');
const path = require('path');
const puppeteer = require('puppeteer-core');

const log = [];
const note = (m) => { log.push(m); console.error('[suno_generate] ' + m); };

const DEFAULT_CHROME = '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome';
const DEFAULT_SESSION = path.join(os.homedir(),
  '.hermes/skills/devops/suno-browser-automation/references/suno-session-persist.json');
// Dedicated, isolated profile — NOT the posting-flow browser. This is what keeps
// Suno generation from contending/locking with posting automation.
const DEFAULT_PROFILE = path.join(os.homedir(), '.hermes/production/.suno-gen-profile');

function parseArgs() {
  let raw = process.argv[2] || '{}';
  let a;
  try { a = JSON.parse(raw); } catch (e) { a = {}; }
  return {
    prompt: a.prompt || '',
    instrumental: a.instrumental !== false,        // default true
    dryRun: !!a.dryRun,
    timeoutMs: a.timeoutMs || 240000,              // 4 min for generation
    expectClips: a.expectClips || 2,               // Suno makes 2 per request
    // Browser sourcing:
    cdpUrl: a.cdpUrl || process.env.SUNO_CDP_URL || '',  // empty => launch our own
    sessionFile: a.sessionFile || process.env.SUNO_SESSION_FILE || DEFAULT_SESSION,
    userDataDir: a.userDataDir || DEFAULT_PROFILE,
    chromePath: a.chromePath || process.env.SUNO_CHROME_PATH || DEFAULT_CHROME,
    headless: a.headless !== false,                // default headless (isolated, no UI)
  };
}

// Load the engagement bot's persisted Suno session (a flat name->value cookie
// map) and inject it so the dedicated browser is logged in — no manual login.
async function injectSession(page, sessionFile) {
  if (!sessionFile || !fs.existsSync(sessionFile)) {
    note(`session file not found: ${sessionFile} (continuing; may not be logged in)`);
    return 0;
  }
  let data;
  try { data = JSON.parse(fs.readFileSync(sessionFile, 'utf8')); }
  catch (e) { note('session file unreadable: ' + e.message); return 0; }
  const META = new Set(['captured_at', '_refreshed_note']);
  const cookies = [];
  for (const [name, value] of Object.entries(data)) {
    if (META.has(name) || typeof value !== 'string') continue;
    // Set on the apex so it covers suno.com + studio-api-prod.suno.com.
    cookies.push({ name, value, domain: '.suno.com', path: '/' });
  }
  try { await page.setCookie(...cookies); note(`injected ${cookies.length} session cookies`); }
  catch (e) { note('setCookie failed: ' + e.message); }
  return cookies.length;
}

// List this account's recent clips via the clean API. Confirmed shape:
//   GET https://studio-api-prod.suno.com/api/project/default
//   -> { project_clips: [ { clip: { id, title, status, audio_url, ... } } ] }
// Auth = Clerk Bearer token (window.Clerk.session.getToken()), NOT just cookies,
// and the base is studio-api-prod.suno.com (relative /api/* hits the wrong origin).
async function recentClips(page) {
  return page.evaluate(async () => {
    const API = 'https://studio-api-prod.suno.com';
    let tok = null;
    try { tok = (window.Clerk && window.Clerk.session) ? await window.Clerk.session.getToken() : null; } catch (e) {}
    try {
      const r = await fetch(API + '/api/project/default', {
        headers: tok ? { Authorization: 'Bearer ' + tok } : {}, credentials: 'include',
      });
      if (!r.ok) return { __status: r.status };
      const j = await r.json();
      const clips = (j.project_clips || []).map((pc) => pc.clip).filter(Boolean).map((c) => ({
        id: c.id, title: c.title || '', status: c.status || '',
        audioUrl: c.audio_url || '', imageUrl: c.image_url || '',
        duration: (c.metadata && (c.metadata.duration || c.metadata.duration_seconds)) || null,
      }));
      return { clips };
    } catch (e) { return { __error: String(e).slice(0, 120) }; }
  });
}

// React controlled inputs ignore JS-injected values — use REAL keystrokes so the
// app's state updates and the Create button enables.
async function setDescription(page, text) {
  // ALL the create-panel placeholders are RANDOMIZED examples, so placeholder
  // matching is unreliable. Instead, find the textarea inside the box headed
  // "Song Description" (the stable label). Verified via screenshot.
  await page.keyboard.press('Escape').catch(() => {});  // dismiss any stray modal
  const handle = await page.evaluateHandle(() => {
    const heads = Array.from(document.querySelectorAll('*'))
      .filter((e) => /^song description$/i.test((e.textContent || '').trim()) && e.children.length <= 1);
    for (const hd of heads) {
      let box = hd;
      for (let i = 0; i < 6 && box; i++) {
        const ta = box.querySelector && box.querySelector('textarea');
        if (ta) return ta;
        box = box.parentElement;
      }
    }
    return Array.from(document.querySelectorAll('textarea')).find((t) => t.offsetParent !== null) || null;
  });
  const el = handle.asElement();
  if (!el) return false;
  await el.click().catch(() => {});
  await page.keyboard.down('Meta').catch(() => {});
  await page.keyboard.press('a').catch(() => {});
  await page.keyboard.up('Meta').catch(() => {});
  await page.keyboard.press('Backspace').catch(() => {});    // clear any existing text
  await el.type(text, { delay: 12 });                         // real keystrokes
  return true;
}

// Locate the generate button (the prominent "Create", not the sidebar nav link).
async function createButton(page) {
  const h = await page.evaluateHandle(() => {
    const btns = Array.from(document.querySelectorAll('button, [role="button"]'));
    // exact "Create" / "Create song" wins; nav link is an <a>, so excluded.
    return btns.find((b) => /^create( song)?$/i.test((b.textContent || '').trim()))
        || btns.find((b) => /create/i.test((b.getAttribute('aria-label') || ''))) || null;
  });
  return h.asElement();
}

async function createButtonState(page) {
  const btn = await createButton(page);
  if (!btn) return { found: false };
  // Only trust real disabled signals — NOT className (Tailwind keeps `disabled:`
  // variant classes in the string at all times, which would false-positive).
  const disabled = await btn.evaluate((b) => b.disabled === true
    || b.getAttribute('aria-disabled') === 'true');
  return { found: true, disabled };
}

async function ensureInstrumental(page, want) {
  return page.evaluate((on) => {
    const els = Array.from(document.querySelectorAll('button, [role="checkbox"], [role="switch"]'));
    const el = els.find((b) => /instrumental/i.test(
      (b.textContent || '') + ' ' + (b.getAttribute('aria-label') || '')));
    if (!el) return 'not-found';
    const checked = el.getAttribute('aria-checked') === 'true'
      || /(^|\s)(checked|active|selected)(\s|$)/i.test(el.className);
    if (checked !== on) { el.click(); return on ? 'enabled' : 'disabled'; }
    return 'already-' + (on ? 'on' : 'off');
  }, want);
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function main() {
  const args = parseArgs();
  if (!args.prompt) {
    console.log(JSON.stringify({ ok: false, error: 'prompt required', log }));
    return;
  }

  let browser;
  let launched = false;
  try {
    if (args.cdpUrl) {
      note(`connecting to existing Chrome at ${args.cdpUrl}`);
      browser = await puppeteer.connect({ browserURL: args.cdpUrl, defaultViewport: null });
    } else {
      note(`launching dedicated ${args.headless ? 'headless ' : ''}Chrome (isolated profile: ${args.userDataDir})`);
      browser = await puppeteer.launch({
        executablePath: args.chromePath,
        headless: args.headless ? 'new' : false,
        userDataDir: args.userDataDir,
        args: ['--no-first-run', '--no-default-browser-check', '--disable-blink-features=AutomationControlled'],
      });
      launched = true;
    }
  } catch (e) {
    console.log(JSON.stringify({
      ok: false,
      error: (args.cdpUrl
        ? `cannot connect to Chrome at ${args.cdpUrl}: ${e.message}`
        : `cannot launch Chrome at ${args.chromePath}: ${e.message}. Set chromePath/SUNO_CHROME_PATH.`),
      log,
    }));
    return;
  }

  const finish = async (obj) => {
    try { launched ? await browser.close() : await browser.disconnect(); } catch (_) {}
    console.log(JSON.stringify(obj));
  };

  try {
    const pages = await browser.pages();
    let page = pages.find((p) => /suno\.com/.test(p.url())) || (await browser.newPage());
    await page.bringToFront().catch(() => {});
    // Inject persisted session BEFORE loading suno.com so we land logged in.
    await injectSession(page, args.sessionFile);
    if (!/suno\.com\/create/.test(page.url())) {
      note('navigating to suno.com/create');
      await page.goto('https://suno.com/create', { waitUntil: 'domcontentloaded', timeout: 60000 });
    } else {
      await page.reload({ waitUntil: 'domcontentloaded', timeout: 60000 }).catch(() => {});
    }
    await sleep(2500);

    // Login check: the create textarea must exist.
    const ok = await setDescription(page, args.prompt);
    if (!ok) {
      await finish({
        ok: false,
        error: 'create UI not found — session likely not valid/logged in. ' +
               'Refresh the persisted Suno session (sessionFile).',
        log,
      });
      return;
    }
    note('description set');
    const instr = await ensureInstrumental(page, args.instrumental);
    note(`instrumental: ${instr}`);

    const beforeResp = await recentClips(page);
    const before = (beforeResp.clips || []).map((c) => c.id);
    note(`existing clips before: ${before.length}` + (beforeResp.__status ? ` (api ${beforeResp.__status})` : ''));

    // React enables Create a beat after the prompt registers — poll for it.
    let btnState = await createButtonState(page);
    for (let i = 0; i < 8 && btnState.found && btnState.disabled; i++) {
      await sleep(1000);
      btnState = await createButtonState(page);
    }
    note(`create button: ${JSON.stringify(btnState)}`);

    if (args.dryRun) {
      await finish({
        ok: true, dryRun: true,
        plan: { prompt: args.prompt, instrumental: args.instrumental, wouldClick: 'Create' },
        createButton: btnState, existingClips: before.length, log,
      });
      return;
    }

    if (!btnState.found) {
      await finish({ ok: false, error: 'Create button not found', log });
      return;
    }
    if (btnState.disabled) {
      await finish({ ok: false, error: 'Create button disabled (prompt not registered?)', log });
      return;
    }
    const btn = await createButton(page);
    await btn.click();                                  // real mouse click
    note('clicked Create');

    // Poll the account's recent clips for the new ones reaching `complete` with audio.
    const beforeSet = new Set(before);
    const deadline = Date.now() + args.timeoutMs;
    let ready = [];
    while (Date.now() < deadline) {
      await sleep(6000);
      const resp = await recentClips(page);
      const fresh = (resp.clips || []).filter((c) => !beforeSet.has(c.id));
      const done = fresh.filter((c) => c.audioUrl && (c.status === 'complete' || c.status === 'streaming'));
      note(`fresh=${fresh.length} ready=${done.length} statuses=[${fresh.map((c) => c.status).join(',')}]`);
      if (done.length >= Math.min(args.expectClips, 1)) { ready = done.slice(0, args.expectClips); break; }
    }

    await finish({
      ok: ready.length > 0,
      clips: ready,
      error: ready.length ? undefined : 'timed out waiting for audio URLs',
      log,
    });
  } catch (e) {
    await finish({ ok: false, error: e.message, log });
  }
}

main();

File: src/suno_generate.py
"""
suno_generate.py — browser-drive Suno music generation for the production pipeline.

Why browser-drive: Suno routes song *generation* through an obfuscated, rotating
service-worker proxy, so it can't be replayed as a clean API call (unlike the
read/social endpoints the engagement bot uses). The robust, non-circumventing
path is to automate a real logged-in browser — click "Create" like a user — then
download the finished MP3 from Suno's public CDN.

Architecture:
    generate (browser)            poll + read           download
    scripts/suno/suno_generate.js  → /api/feed (in-page) → cdn1.suno.ai/{id}.mp3
    (Node + puppeteer-core, CDP)                            (this module, httpx)

Dry-run is the DEFAULT and does nothing external (no browser, no generation, no
download) — it just prints the plan. Use --execute to actually generate (paid:
consumes Suno credits and creates clips on the logged-in account).

No manual login: --execute launches its OWN dedicated, isolated headless Chrome
(separate profile dir) and injects the engagement bot's persisted Suno session
cookies. Because it's a separate browser instance, it does NOT contend with /
lock the posting-flow browser. (Generation must run in a browser at all because
Suno proxies it through an obfuscated service worker — it can't be a plain curl
call like the engagement bot's read/social actions.)

CLI:
    python -m src.suno_generate --prompt "dark cinematic ambient, 528hz"          # dry-run
    python -m src.suno_generate --prompt "..." --execute --out-dir /tmp/suno      # real
    python -m src.suno_generate --prompt "..." --execute --headful               # show the browser

Library:
    from src.suno_generate import generate_song
    result = generate_song("dark cinematic ambient", instrumental=True, dry_run=False)
    result.audio_paths  # -> [Path(...mp3), ...]
"""

from __future__ import annotations

import argparse
import json
import logging
import subprocess
import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional

logger = logging.getLogger(__name__)

REPO_ROOT = Path(__file__).resolve().parent.parent
NODE_SCRIPT = REPO_ROOT / "scripts" / "suno" / "suno_generate.js"
DEFAULT_OUT_DIR = Path.home() / ".hermes" / "production" / "clips" / "audio"
DEFAULT_SESSION = (Path.home() / ".hermes" / "skills" / "devops"
                   / "suno-browser-automation" / "references" / "suno-session-persist.json")
DEFAULT_PROFILE = Path.home() / ".hermes" / "production" / ".suno-gen-profile"
DEFAULT_CHROME = "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"
DEFAULT_TIMEOUT_S = 300


@dataclass
class SunoClip:
    id: str
    title: str
    status: str
    audio_url: str
    duration: Optional[float] = None
    audio_path: Optional[Path] = None


@dataclass
class SunoResult:
    ok: bool
    dry_run: bool
    prompt: str
    instrumental: bool
    clips: list[SunoClip] = field(default_factory=list)
    error: Optional[str] = None

    @property
    def audio_paths(self) -> list[Path]:
        return [c.audio_path for c in self.clips if c.audio_path]

    def to_dict(self) -> dict:
        """Stable contract for callers (e.g. the production audio_layer bridge)."""
        first = self.clips[0] if self.clips else None
        return {
            "ok": self.ok,
            "dry_run": self.dry_run,
            "prompt": self.prompt,
            "instrumental": self.instrumental,
            "error": self.error,
            # convenience fields (first clip)
            "title": first.title if first else None,
            "status": first.status if first else None,
            "duration": first.duration if first else None,
            "paths": [str(p) for p in self.audio_paths],
            # full per-clip detail
            "clips": [
                {"id": c.id, "title": c.title, "status": c.status,
                 "duration": c.duration, "path": str(c.audio_path) if c.audio_path else None}
                for c in self.clips
            ],
        }


def generate_song(
    prompt: str,
    *,
    instrumental: bool = True,
    dry_run: bool = True,
    out_dir: Path = DEFAULT_OUT_DIR,
    timeout_s: int = DEFAULT_TIMEOUT_S,
    session_file: Path = DEFAULT_SESSION,
    user_data_dir: Path = DEFAULT_PROFILE,
    chrome_path: str = DEFAULT_CHROME,
    headless: bool = True,
    cdp_url: str = "",
    node_bin: str = "node",
) -> SunoResult:
    """Generate a Suno song via an isolated browser and download the MP3(s).

    Launches its own dedicated Chrome (own profile) and injects the persisted
    Suno session — no login, no contention with the posting-flow browser. If
    cdp_url is given, attaches to that running Chrome instead of launching.

    Dry-run (default) performs NO external action — it returns a planned result.
    """
    prompt = (prompt or "").strip()
    if not prompt:
        return SunoResult(False, dry_run, prompt, instrumental, error="empty prompt")

    if dry_run:
        logger.info("[suno] DRY RUN — would generate: %r (instrumental=%s)", prompt, instrumental)
        return SunoResult(True, True, prompt, instrumental)

    if not NODE_SCRIPT.exists():
        return SunoResult(False, False, prompt, instrumental,
                          error=f"node script missing: {NODE_SCRIPT}")

    node_args = {
        "prompt": prompt,
        "instrumental": instrumental,
        "dryRun": False,
        "timeoutMs": timeout_s * 1000,
        "sessionFile": str(session_file),
        "userDataDir": str(user_data_dir),
        "chromePath": chrome_path,
        "headless": headless,
    }
    if cdp_url:
        node_args["cdpUrl"] = cdp_url
    try:
        proc = subprocess.run(
            [node_bin, str(NODE_SCRIPT), json.dumps(node_args)],
            capture_output=True, text=True, timeout=timeout_s + 60,
        )
    except subprocess.TimeoutExpired:
        return SunoResult(False, False, prompt, instrumental, error="node process timed out")
    except FileNotFoundError:
        return SunoResult(False, False, prompt, instrumental,
                          error=f"'{node_bin}' not found — is Node installed?")

    payload = _last_json_line(proc.stdout)
    if payload is None:
        return SunoResult(False, False, prompt, instrumental,
                          error=f"no JSON from node. stderr tail: {proc.stderr[-400:]}")

    if not payload.get("ok"):
        return SunoResult(False, False, prompt, instrumental,
                          error=payload.get("error", "generation failed"))

    clips = [
        SunoClip(id=c.get("id", ""), title=c.get("title", ""),
                 status=c.get("status", ""), audio_url=c.get("audioUrl", ""),
                 duration=c.get("duration"))
        for c in payload.get("clips", [])
    ]
    out_dir.mkdir(parents=True, exist_ok=True)
    for clip in clips:
        if clip.audio_url:
            clip.audio_path = _download(clip.audio_url, out_dir / f"{clip.id or 'suno'}.mp3")

    return SunoResult(True, False, prompt, instrumental, clips=clips)


def _last_json_line(stdout: str) -> Optional[dict]:
    for line in reversed((stdout or "").strip().splitlines()):
        line = line.strip()
        if line.startswith("{"):
            try:
                return json.loads(line)
            except json.JSONDecodeError:
                continue
    return None


def _download(url: str, dest: Path) -> Optional[Path]:
    """Download a CDN MP3 to dest. Suno audio lives on public cdn1.suno.ai."""
    try:
        import httpx
        with httpx.stream("GET", url, follow_redirects=True, timeout=120) as r:
            r.raise_for_status()
            with open(dest, "wb") as f:
                for chunk in r.iter_bytes():
                    f.write(chunk)
        logger.info("[suno] downloaded %s", dest)
        return dest
    except Exception as e:  # noqa: BLE001 - download is best-effort; report and continue
        logger.warning("[suno] download failed for %s: %s", url, e)
        return None


def main(argv: Optional[list[str]] = None) -> int:
    p = argparse.ArgumentParser(description="Browser-drive Suno music generation (dry-run by default).")
    p.add_argument("--prompt", required=True, help="Song description / style prompt.")
    p.add_argument("--instrumental", dest="instrumental", action="store_true", default=True,
                   help="Generate instrumental only (default).")
    p.add_argument("--no-instrumental", dest="instrumental", action="store_false",
                   help="Allow vocals.")
    p.add_argument("--execute", action="store_true",
                   help="Actually generate (paid: uses credits, creates clips). Default: dry-run.")
    p.add_argument("--out-dir", type=Path, default=DEFAULT_OUT_DIR, help="Where to save MP3s.")
    p.add_argument("--timeout", type=int, default=DEFAULT_TIMEOUT_S, help="Generation timeout (s).")
    p.add_argument("--session-file", type=Path, default=DEFAULT_SESSION,
                   help="Persisted Suno session cookies to inject (no manual login).")
    p.add_argument("--user-data-dir", type=Path, default=DEFAULT_PROFILE,
                   help="Dedicated, isolated Chrome profile (kept separate from posting flows).")
    p.add_argument("--chrome-path", default=DEFAULT_CHROME, help="Path to the Chrome binary.")
    p.add_argument("--headful", dest="headless", action="store_false", default=True,
                   help="Show the browser window (default: headless).")
    p.add_argument("--cdp-url", default="",
                   help="Attach to an already-running Chrome instead of launching one.")
    p.add_argument("--json", action="store_true",
                   help="Emit a single JSON result line (stable contract for callers like audio_layer).")
    args = p.parse_args(argv)

    logging.basicConfig(level=logging.WARNING if args.json else logging.INFO, format="%(message)s")
    dry_run = not args.execute

    if args.json:
        result = generate_song(
            args.prompt, instrumental=args.instrumental, dry_run=dry_run,
            out_dir=args.out_dir, timeout_s=args.timeout,
            session_file=args.session_file, user_data_dir=args.user_data_dir,
            chrome_path=args.chrome_path, headless=args.headless, cdp_url=args.cdp_url,
        )
        print(json.dumps(result.to_dict()))
        return 0 if result.ok else 1

    print("Suno music generation (browser-drive)")
    print(f"  mode ......... {'EXECUTE (paid)' if not dry_run else 'dry-run'}")
    print(f"  prompt ....... {args.prompt!r}")
    print(f"  instrumental . {args.instrumental}")
    if not dry_run:
        print(f"  browser ...... {'attach ' + args.cdp_url if args.cdp_url else ('launch ' + ('headless' if args.headless else 'headful'))}")
        print(f"  session ...... {args.session_file}")
        print(f"  profile ...... {args.user_data_dir}  (isolated from posting flows)")
        print(f"  out dir ...... {args.out_dir}")

    result = generate_song(
        args.prompt, instrumental=args.instrumental, dry_run=dry_run,
        out_dir=args.out_dir, timeout_s=args.timeout,
        session_file=args.session_file, user_data_dir=args.user_data_dir,
        chrome_path=args.chrome_path, headless=args.headless, cdp_url=args.cdp_url,
    )

    print("─" * 56)
    if result.dry_run:
        print("  DRY RUN — no browser, no generation, no download.")
        print("  Re-run with --execute to generate for real (launches its own")
        print("  isolated headless Chrome + injects the persisted Suno session).")
        print("─" * 56)
        return 0

    if not result.ok:
        print(f"  FAILED: {result.error}")
        print("─" * 56)
        return 1

    print(f"  Generated {len(result.clips)} clip(s):")
    for c in result.clips:
        loc = c.audio_path if c.audio_path else "(download failed)"
        print(f"    • {c.title or c.id} [{c.status}] → {loc}")
    print("─" * 56)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
